minio/api.py

```diff
@@ -615,58 +674,28 @@ def fput_object(self, bucket_name, object_name, file_path,
 
-        # Always start with first part number.
+        # Threads pool which upload parts
+        p = multiprocessing.dummy.Pool(_PARALLEL_UPLOADERS)
+
+        parts_to_upload = []
+
+        # Build list of parts to upload, always start with first part number.
         for part_number in range(1, total_parts_count + 1):
-            # Save the current part size that needs to be uploaded.
+            # Calculate the current part size
             current_part_size = part_size
             if part_number == total_parts_count:
                 current_part_size = last_part_size
+            # Append current part information
+            parts_to_upload.append((uploaded_parts, file_path, bucket_name, object_name,
+                upload_id, part_number, current_part_size)) 
 
-            # Save current offset as previous offset.
-            prev_offset = file_data.seek(0, 1)
-
-            # Calculate md5sum and sha256.
-            md5hasher = Hasher.md5()
-            sha256hasher = Hasher.sha256()
-            total_read = 0
-
-            # Save LimitedReader, read upto current_part_size for
-            # md5sum and sha256 calculation.
-            part = LimitedReader(file_data, current_part_size)
-            while total_read < current_part_size:
-                current_data = part.read()  # Read in 64k chunks.
-                if not current_data or len(current_data) == 0:
-                    break
-                md5hasher.update(current_data)
-                sha256hasher.update(current_data)
-                total_read += len(current_data)
-
-            part_md5_hex = md5hasher.hexdigest()
-            # Verify if current part number has been already
-            # uploaded. Verify if the size is same, further verify if
-            # we have matching md5sum as well.
-            if part_number in uploaded_parts:
-                previous_part = uploaded_parts[part_number]
-                if previous_part.size == current_part_size:
-                    if previous_part.etag == part_md5_hex:
-                        total_uploaded += previous_part.size
-                        continue
-
-            # Seek back to previous offset position before checksum
-            # calculation.
-            file_data.seek(prev_offset, 0)
-
-            # Create the LimitedReader again for the http reader.
-            part = LimitedReader(file_data, current_part_size)
-            part_metadata = PartMetadata(part, md5hasher.hexdigest(), sha256hasher.hexdigest(),
-                                         current_part_size)
-            # Initiate multipart put.
-            etag = self._do_put_multipart_object(bucket_name, object_name,
-                                                 part_metadata,
-                                                 upload_id,
-                                                 part_number)
+        # Run part uploads in parallel
+        part_uploads_result = p.map(self._upload_part, parts_to_upload)
```

Comment:
@donatello, it is possible to catch the exception from the code calling our fput_object. I made a typo and I didn't saw it. Multiprocessing stops sending work to the threads pool when an uncaught exception is raised. 

However, I noticed another problem which is that multiprocessor in python 2 doesn't detect keyboard interruption (bug) but python 3 does. I might have to change everything due to this issue..